### PR TITLE
cleaner training for dynamic PR

### PR DESCRIPTION
**Pull request by Zenglinxiao**
This PR propose cleaner training by:
- Initialize necessary stuff for training (fields/checkpoints/transforms) before running `train_single` on each GPU;
- Pass these items to each training replica rather than initialize them in each process;
- Avoid dump fields/transforms, make it optional.

### CODE REVIEW

File: onmt/bin/train.py [hunk 1]
```diff
@@ -24,38 +26,85 @@
...
         save_transformed_sample(opt, transforms, n_sample=opt.n_sample)
-        sys.exit("Sample saved, please check it before restart training.")
+        logger.info(
+            "Sample saved, please check it before restart training.")
+        sys.exit()
+    return fields, transforms_cls
+
+
+def _init_train(opt):
+    """Common initilization stuff for all training replica."""
```
Comment (francoishernandez) [inline onmt/bin/train.py]: Not sure "replica" is very explicit.
Maybe "process" / "subprocess" / "multiprocessing process" / "thread" ? (not sure which one)

File: onmt/bin/train.py [hunk 2]
```diff
@@ -24,38 +26,85 @@
...
+    DynamicArgumentParser.validate_prepare_opts(opt)
+
+    if opt.train_from:
+        # Load checkpoint if we resume from a previous training.
+        checkpoint = load_checkpoint(ckpt_path=opt.train_from)
+        fields = load_fields(opt.save_data, checkpoint)
+        transforms_cls = get_transforms_cls(opt._all_transform)
+        if (hasattr(checkpoint["opt"], '_all_transform') and
```
Comment (francoishernandez) [inline onmt/bin/train.py:64]: This check is a bit strange, since the line above assumes it exists anyways.

File: onmt/bin/train.py [hunk 3]
```diff
@@ -24,38 +26,85 @@
...
     DynamicArgumentParser.update_model_opts(opt)
     DynamicArgumentParser.validate_model_opts(opt)
 
-    init_logger(opt.log_file)
     set_random_seed(opt.seed, False)
 
-    if not opt.train_from:
-        prepare_fields_transforms(opt)
+    checkpoint, fields, transforms_cls = _init_train(opt)
+    train_replica = partial(
```
Comment (francoishernandez) [inline onmt/bin/train.py]: same remark about replica

File: onmt/dynamic/opts.py [hunk 4]
```diff
@@ -24,6 +22,23 @@ def _dynamic_corpus_opts(parser):
...
+        '-n_sample', '--n_sample',
+        type=int, default=(5000 if build_vocab_only else 0),
+        help=("Build vocab using " if build_vocab_only else "Stop after save ")
+        + "this number of transformed samples/corpus. Can be [-1, 0, N>0]. "
+        "Set to -1 to go full corpus, 0 to skip.")
+
+    if not build_vocab_only:
+        group.add('-dump_fields', '--dump_fields', action='store_true',
+                  help="Dump fields `*.vocab.pt` to disk."
+                  " -save_data should be set as saving prefix.")
```
Comment (francoishernandez) [inline onmt/dynamic/opts.py:37]: Wouldn't it be more straightforward if `-dump_fields` was the path where we want to save it, not only a boolean flag?
(same for `-dump_transforms`)

File: onmt/models/model_saver.py [hunk 5]
```diff
@@ -72,11 +82,12 @@ def save(self, step, moving_average=None):
                 self._rm_checkpoint(todel)
             self.checkpoint_queue.append(chkpt_name)
 
-    def _save(self, step):
+    def _save(self, step, model):
```
Comment (francoishernandez) [inline onmt/models/model_saver.py:85]: Why this change here?

File: onmt/dynamic/opts.py (hunk 4 above)
Comment (francoishernandez) [reply onmt/dynamic/opts.py:37]: not sure about embeddings though

File: onmt/models/model_saver.py (hunk 5 above)
Comment (Zenglinxiao) [reply onmt/models/model_saver.py:85]: It's not defined properly in the base class. We call `self._save(step, model)` in `self.save` method.

File: onmt/dynamic/opts.py (hunk 4 above)
Comment (Zenglinxiao) [reply onmt/dynamic/opts.py:37]: Use a single `save_data` to control the saving path for `fields/transforms/samples/pretrained embeddings` seems cleaner to me rather than give one path for each which needs copy/paste in most cases.

File: onmt/bin/train.py (hunk 2 above)
Comment (Zenglinxiao) [reply onmt/bin/train.py:64]: It's to avoid failing when `-train_from` an old model (which doesn't have `_all_transforms` in the checkpoint's opt).

File: onmt/bin/train.py (hunk 1 above)
Comment (Zenglinxiao) [reply onmt/bin/train.py]: I'll take `process` I think.

File: onmt/dynamic/opts.py (hunk 4 above)
Comment (francoishernandez) [reply onmt/dynamic/opts.py:37]: I think both approaches are valid. But alright let's keep `save_data` for now.

File: onmt/bin/train.py (hunk 2 above)
Comment (francoishernandez) [reply onmt/bin/train.py:64]: I understand that. But why is it not performed for the line above also?
